proof/refine/X64/Tcb_R.thy

```diff
@@ -1704,66 +1704,60 @@ crunch set_flags
   and pspace_distinct[wp]: pspace_distinct
   and valid_cur_fpu[wp]: valid_cur_fpu
 
-lemma tcbFlagToWord_bit:
-  "\<exists>n. tcbFlagToWord flag = bit n"
-  by (auto simp: tcbFlagToWord_def split: tcb_flag.splits simp del: bit_def)
-
-lemma tcbFlagToWord_le_mask_numFlags:
-  "tcbFlagToWord flag \<le> mask numFlags"
-  by (auto simp: tcbFlagToWord_def numFlags_def split: tcb_flag.splits)
+(*FIXME FPU: move tcbFlag lemmas*)
+lemma word_to_tcb_flags_not:
+  "word_to_tcb_flags (~~ flags) = word_to_tcb_flags tcbFlagMask - word_to_tcb_flags flags"
+  apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
+  apply (cut_tac flag=x in tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
+  done
 
-\<comment> \<open>This will probably be much harder if there are more flags\<close>
-lemma ex_tcbFlagToWord:
-  "n < numFlags \<Longrightarrow> \<exists>flag. tcbFlagToWord flag = bit n"
-  by (auto simp: tcbFlagToWord_def numFlags_def split: tcb_flag.splits)
+lemma word_to_tcb_flags_or:
+  "word_to_tcb_flags (flags || flags') = word_to_tcb_flags flags \<union> word_to_tcb_flags flags'"
+  apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
+  apply (cut_tac flag=x in tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
+  done
 
-lemma word_to_tcb_flags_subtract:
-  "word_to_tcb_flags (flags && ~~ flags') = word_to_tcb_flags flags - word_to_tcb_flags flags'"
+lemma word_to_tcb_flags_and:
+  "word_to_tcb_flags (flags && flags') = word_to_tcb_flags flags \<inter> word_to_tcb_flags flags'"
   apply (clarsimp simp: word_to_tcb_flags_def intro!: set_eqI)
   apply (cut_tac flag=x in tcbFlagToWord_bit)
-  apply (fastforce simp: word_eq_iff tcbFlagToWord_bit)
+  apply (fastforce simp: word_eq_iff)
   done
 
-lemma word_to_tcb_flags_union:
-  "word_to_tcb_flags (flags || flags') = word_to_tcb_flags flags \<union> word_to_tcb_flags flags'"
+lemmas word_to_tcb_flags_simps = word_to_tcb_flags_not word_to_tcb_flags_or word_to_tcb_flags_and
```

Comment:
reading through these a few times (once per arch), I keep thinking: are these properties arch-agnostic? maybe it'll have to be absorbed in arch-split if so, but it really looks like some kind of library that could even be spun out into its own little theory

Reply:
You're right, these properties are arch-agnostic now. The proofs for them weren't originally, and them being so now is one of the things giving me some confidence that they'll be able to be extended for additional flags in the future.

I'd been thinking along the lines of them being absorbed in arch-split. The exact form of some of these lemmas is pretty specific to how they're used and it can be annoyingly fiddly to get the simplifier to apply some of them due to fighting with associativity and commutativity, so spinning them out on their own would take a bit of work to do nicely. That might be something to come back to but I'd prefer not as part of this PR.